File: archive/navvis-cleanroom-alignment/code/src/navvis_recon/image_postprocessing.py
"""DNG/image post-processing reference path.

The binary imports LibRaw/DNG SDK and OpenCV's MergeMertens, GaussianBlur,
medianBlur, fastNlMeansDenoising, addWeighted and JPEG encoder.  Those observed
operations define the pipeline below; raw decoding itself remains delegated to
the caller.
"""

from __future__ import annotations

from dataclasses import dataclass
from pathlib import Path
from typing import Iterable, Sequence

import cv2
import numpy as np


@dataclass(slots=True)
class ImageConfig:
    preset: str = "high-quality"  # high-quality | fast | plain
    sharpen: float = 1.0
    denoise: str | float = -1.0  # adaptive wavelet, 0/off, positive/manual, nlm, nlm-fused
    exposure_stops: float = 0.0
    white_balance: str = "camera"  # camera | auto | custom | per-pano
    custom_white_balance: tuple[float, float, float] = (1.0, 1.0, 1.0)
    hdr_ev_shifts: tuple[float, ...] = (-1.0, 0.0, 1.0)
    max_file_size: int | None = None


def gray_world_white_balance(linear_rgb: np.ndarray) -> np.ndarray:
    means = np.maximum(linear_rgb.reshape(-1, 3).mean(axis=0), 1e-6)
    gains = means.mean() / means
    return np.clip(linear_rgb * gains[None, None, :], 0, 1)


def apply_vignetting(image: np.ndarray, polynomial: Sequence[float]) -> np.ndarray:
    """Apply an inverse radial vignetting polynomial centered in the image."""
    h, w = image.shape[:2]
    yy, xx = np.mgrid[:h, :w]
    x = (xx - (w - 1) / 2) / max(w / 2, 1)
    y = (yy - (h - 1) / 2) / max(h / 2, 1)
    r2 = x * x + y * y
    attenuation = np.zeros_like(r2, dtype=float)
    power = np.ones_like(r2, dtype=float)
    for coefficient in polynomial:
        attenuation += coefficient * power
        power *= r2
    return np.clip(image / np.maximum(attenuation[..., None], 0.1), 0, 1)


def exposure_fusion(linear_rgb: np.ndarray, ev_shifts: Iterable[float]) -> np.ndarray:
    exposures = []
    for ev in ev_shifts:
        shifted = np.clip(linear_rgb * (2.0 ** ev), 0, 1)
        srgb = np.where(shifted <= 0.0031308, 12.92 * shifted, 1.055 * shifted ** (1 / 2.4) - 0.055)
        exposures.append(np.rint(255 * srgb[..., ::-1]).astype(np.uint8))
    fused = cv2.createMergeMertens().process(exposures)
    return np.clip(fused, 0, 1)


def denoise(image_bgr: np.ndarray, mode: str | float) -> np.ndarray:
    if mode in (0, 0.0, "0", "off"):
        return image_bgr
    image8 = np.rint(np.clip(image_bgr, 0, 1) * 255).astype(np.uint8)
    if mode in ("nlm", "nlm-fused"):
        h = 5 if mode == "nlm-fused" else 7
        return cv2.fastNlMeansDenoisingColored(image8, None, h, h, 7, 21).astype(np.float32) / 255
    strength = 1.0 if float(mode) < 0 else float(mode)
    # Edge-preserving wavelet equivalent: subtract a bilateral base and shrink
    # small high-frequency coefficients.
    base = cv2.bilateralFilter(image8, 7, 25, 5).astype(np.float32) / 255
    detail = image_bgr - base
    threshold = 0.008 * strength
    detail = np.sign(detail) * np.maximum(np.abs(detail) - threshold, 0)
    return np.clip(base + detail, 0, 1)


def unsharp_mask(image: np.ndarray, amount: float) -> np.ndarray:
    if amount <= 0:
        return image
    blurred = cv2.GaussianBlur(image, (0, 0), 1.2)
    return np.clip(cv2.addWeighted(image, 1.0 + amount, blurred, -amount, 0), 0, 1)


def blur_regions(image: np.ndarray, regions: Iterable[tuple[int, int, int, int]], kernel: int = 31) -> np.ndarray:
    result = image.copy()
    kernel = max(3, kernel | 1)
    for x, y, w, h in regions:
        x0, y0 = max(0, x), max(0, y)
        x1, y1 = min(image.shape[1], x + w), min(image.shape[0], y + h)
        if x1 > x0 and y1 > y0:
            result[y0:y1, x0:x1] = cv2.GaussianBlur(result[y0:y1, x0:x1], (kernel, kernel), 0)
    return result


def process_linear_rgb(
    linear_rgb: np.ndarray,
    cfg: ImageConfig,
    camera_white_balance: tuple[float, float, float] = (1, 1, 1),
    vignetting: Sequence[float] = (1.0,),
    blur_boxes: Iterable[tuple[int, int, int, int]] = (),
) -> np.ndarray:
    image = np.asarray(linear_rgb, dtype=np.float32)
    if cfg.white_balance == "auto":
        image = gray_world_white_balance(image)
    else:
        gains = cfg.custom_white_balance if cfg.white_balance == "custom" else camera_white_balance
        image = np.clip(image * np.asarray(gains)[None, None, :], 0, 1)
    image = apply_vignetting(image, vignetting)
    image = np.clip(image * (2.0 ** cfg.exposure_stops), 0, 1)
    if cfg.preset == "plain":
        image_bgr = image[..., ::-1]
    else:
        shifts = (0.0,) if cfg.preset == "fast" else cfg.hdr_ev_shifts
        image_bgr = exposure_fusion(image, shifts)
        image_bgr = denoise(image_bgr, cfg.denoise)
        image_bgr = unsharp_mask(image_bgr, cfg.sharpen)
    return blur_regions(image_bgr, blur_boxes)


def encode_jpeg(image_bgr: np.ndarray, max_bytes: int | None = None) -> bytes:
    image8 = np.rint(np.clip(image_bgr, 0, 1) * 255).astype(np.uint8)
    qualities = range(95, 34, -5) if max_bytes else (95,)
    last = b""
    for quality in qualities:
        ok, encoded = cv2.imencode(".jpg", image8, [cv2.IMWRITE_JPEG_QUALITY, quality])
        if not ok:
            raise RuntimeError("JPEG encoding failed")
        last = encoded.tobytes()
        if max_bytes is None or len(last) <= max_bytes:
            return last
    return last


def write_jpeg(path: str | Path, image_bgr: np.ndarray, cfg: ImageConfig) -> None:
    Path(path).write_bytes(encode_jpeg(image_bgr, cfg.max_file_size))

File: archive/navvis-cleanroom-alignment/code/cpp/src/image_postprocessing.cpp
#include "navvis_recon/image_postprocessing.hpp"

#include <opencv2/imgcodecs.hpp>
#include <opencv2/imgproc.hpp>
#include <opencv2/photo.hpp>

#include <algorithm>
#include <cmath>

namespace navvis_recon {
namespace {

cv::Mat3f clamp01(const cv::Mat3f& input) {
    cv::Mat3f output;
    cv::max(input, 0.0, output);
    cv::min(output, 1.0, output);
    return output;
}

cv::Mat3b to8Bit(const cv::Mat3f& input) {
    cv::Mat3b output;
    clamp01(input).convertTo(output, CV_8UC3, 255.0);
    return output;
}

cv::Mat3f srgbFromLinear(const cv::Mat3f& linear) {
    cv::Mat3f output(linear.size());
    for (int y = 0; y < linear.rows; ++y) {
        for (int x = 0; x < linear.cols; ++x) {
            for (int channel = 0; channel < 3; ++channel) {
                const float value = std::clamp(linear(y, x)[channel], 0.0F, 1.0F);
                output(y, x)[channel] = value <= 0.0031308F
                                              ? 12.92F * value
                                              : 1.055F * std::pow(value, 1.0F / 2.4F) - 0.055F;
            }
        }
    }
    return output;
}

}  // namespace

ImagePostprocessor::ImagePostprocessor(ImagePostprocessingOptions options)
    : options_(std::move(options)) {}

cv::Mat3f ImagePostprocessor::applyWhiteBalance(
    const cv::Mat3f& linear_rgb, ImagePostprocessingOptions::WhiteBalance mode,
    const cv::Vec3f& camera_gains, const cv::Vec3f& custom_gains) {
    cv::Vec3f gains = camera_gains;
    if (mode == ImagePostprocessingOptions::WhiteBalance::Custom) {
        gains = custom_gains;
    } else if (mode == ImagePostprocessingOptions::WhiteBalance::Automatic) {
        cv::Scalar means = cv::mean(linear_rgb);
        const float average = static_cast<float>((means[0] + means[1] + means[2]) / 3.0);
        for (int channel = 0; channel < 3; ++channel) {
            gains[channel] = average / std::max(static_cast<float>(means[channel]), 1.0e-6F);
        }
    }
    cv::Mat3f output = linear_rgb.clone();
    for (int y = 0; y < output.rows; ++y) {
        for (int x = 0; x < output.cols; ++x) {
            output(y, x) = output(y, x).mul(gains);
        }
    }
    return clamp01(output);
}

cv::Mat3f ImagePostprocessor::applyVignetting(
    const cv::Mat3f& image, const std::vector<float>& polynomial) {
    cv::Mat3f output = image.clone();
    const float center_x = 0.5F * static_cast<float>(image.cols - 1);
    const float center_y = 0.5F * static_cast<float>(image.rows - 1);
    for (int y = 0; y < image.rows; ++y) {
        for (int x = 0; x < image.cols; ++x) {
            const float nx = (static_cast<float>(x) - center_x) / std::max(center_x, 1.0F);
            const float ny = (static_cast<float>(y) - center_y) / std::max(center_y, 1.0F);
            const float radius_squared = nx * nx + ny * ny;
            float attenuation = 0.0F;
            float power = 1.0F;
            for (const float coefficient : polynomial) {
                attenuation += coefficient * power;
                power *= radius_squared;
            }
            output(y, x) /= std::max(attenuation, 0.1F);
        }
    }
    return clamp01(output);
}

cv::Mat3f ImagePostprocessor::exposureFusion(
    const cv::Mat3f& linear_rgb, const std::vector<float>& ev_shifts) {
    std::vector<cv::Mat> exposures;
    for (const float shift : ev_shifts) {
        cv::Mat3f shifted = clamp01(linear_rgb * std::pow(2.0F, shift));
        cv::Mat3f srgb = srgbFromLinear(shifted);
        cv::Mat3f bgr;
        cv::cvtColor(srgb, bgr, cv::COLOR_RGB2BGR);
        exposures.push_back(to8Bit(bgr));
    }
    cv::Mat fused;
    cv::createMergeMertens()->process(exposures, fused);
    return clamp01(fused);
}

cv::Mat3f ImagePostprocessor::denoiseImage(
    const cv::Mat3f& image_bgr, ImagePostprocessingOptions::Denoise mode, float strength) {
    if (mode == ImagePostprocessingOptions::Denoise::Off) {
        return image_bgr;
    }
    const cv::Mat3b input8 = to8Bit(image_bgr);
    if (mode == ImagePostprocessingOptions::Denoise::Nlm ||
        mode == ImagePostprocessingOptions::Denoise::NlmFused) {
        cv::Mat3b output8;
        // Reference high-quality path: h=4.6785717, template 7, search 17.
        // Both named NLM modes enter the same OpenCV stage in the binary.
        constexpr float h = 4.6785717F;
        cv::fastNlMeansDenoising(input8, output8, h, 7, 17);
        cv::Mat3f output;
        output8.convertTo(output, CV_32FC3, 1.0 / 255.0);
        return output;
    }
    // The binary names adaptive/manual wavelet modes but strips the private
    // wavelet implementation. This is the equivalent coefficient-shrink path.
    cv::Mat3b base8;
    cv::bilateralFilter(input8, base8, 7, 25.0, 5.0);
    cv::Mat3f base;
    base8.convertTo(base, CV_32FC3, 1.0 / 255.0);
    cv::Mat3f detail = image_bgr - base;
    const float threshold = 0.008F * std::max(strength, 0.1F);
    for (int y = 0; y < detail.rows; ++y) {
        for (int x = 0; x < detail.cols; ++x) {
            for (int channel = 0; channel < 3; ++channel) {
                const float value = detail(y, x)[channel];
                detail(y, x)[channel] = std::copysign(std::max(std::abs(value) - threshold, 0.0F), value);
            }
        }
    }
    return clamp01(base + detail);
}

cv::Mat3f ImagePostprocessor::unsharpMask(const cv::Mat3f& image, float amount) {
    if (amount <= 0.0F) {
        return image;
    }
    cv::Mat3f blurred;
    // The binary forms 1.5 * source - 0.5 * Gaussian(source, sigma=3)
    // for its default amount of 0.5.
    cv::GaussianBlur(image, blurred, cv::Size(), 3.0);
    return clamp01((1.0F + amount) * image - amount * blurred);
}

cv::Mat3f ImagePostprocessor::blurRegions(
    const cv::Mat3f& image, const std::vector<BlurRegion>& regions) {
    cv::Mat3f result = image.clone();
    const cv::Rect bounds(0, 0, image.cols, image.rows);
    for (const auto& region : regions) {
        const cv::Rect clipped = region.rectangle & bounds;
        if (clipped.empty()) {
            continue;
        }
        int kernel = std::max(3, region.kernel_size);
        kernel |= 1;
        cv::GaussianBlur(result(clipped), result(clipped), cv::Size(kernel, kernel), 0.0);
    }
    return result;
}

cv::Mat3f ImagePostprocessor::processLinearRgb(
    const cv::Mat3f& linear_rgb, const cv::Vec3f& camera_white_balance,
    const std::vector<float>& vignetting_polynomial,
    const std::vector<BlurRegion>& blur_regions) const {
    cv::Mat3f image = applyWhiteBalance(
        linear_rgb, options_.white_balance, camera_white_balance, options_.custom_white_balance);
    image = applyVignetting(image, vignetting_polynomial);
    image = clamp01(image * std::pow(2.0F, options_.exposure_stops));
    if (options_.preset == ImagePostprocessingOptions::Preset::Plain) {
        cv::cvtColor(image, image, cv::COLOR_RGB2BGR);
    } else {
        const std::vector<float> shifts =
            options_.preset == ImagePostprocessingOptions::Preset::Fast
                ? std::vector<float>{0.0F}
                : options_.hdr_ev_shifts;
        image = exposureFusion(image, shifts);
        image = denoiseImage(image, options_.denoise, options_.denoise_strength);
        image = unsharpMask(image, options_.sharpen);
    }
    return blurRegions(image, blur_regions);
}

std::vector<std::uint8_t> ImagePostprocessor::encodeJpeg(const cv::Mat3f& image_bgr) const {
    const cv::Mat3b image8 = to8Bit(image_bgr);
    std::vector<std::uint8_t> encoded;
    for (int quality = 95; quality >= 35; quality -= 5) {
        if (!cv::imencode(".jpg", image8, encoded, {cv::IMWRITE_JPEG_QUALITY, quality})) {
            throw std::runtime_error("JPEG encoding failed");
        }
        if (options_.maximum_jpeg_bytes == 0U || encoded.size() <= options_.maximum_jpeg_bytes) {
            return encoded;
        }
    }
    return encoded;
}

}  // namespace navvis_recon
